### PR TITLE
improve(voter-app): update app to work with newest voting version 1.2.2

### LINKED ISSUE #1969: Errors trying to decrypt votes in both voterdapp and cli
**Description**
Unable to decode votes causing votedapp to crash and cli tool to not work on reveal.

**Reproduction**

Build protocol/core and follow directions in voter-dapp/run_tests.sh
Get to snapshot phase 6/10 and enter votes in dapp
see that there is an error in dapp and unable to decode.
Move to reveal phase.
Reveal votes in cli, see that you get a bad decrypt.

**Expected Behavior**

Should just work without errors and show your votes.

**Screenshots**

Using CLI:
image

Using VoterDapp:
image

**Version and OS Info**

Running Ubuntu 18.04. 
Chrome fresh user
Lastest protocol/master
Fresh metamask

**Additional Context**
This might be a linux environment issue, most likely will start to debug there.

### LINKED ISSUE #2371: Re-release existing voter dApp to work with new DVM ABIs
Why
There were breaking changes in the new voting contract, specifically around handling ancillary data. The current implementation is backward compatible currently with the dapp, but we want to update it to call the new functions.

AC
- See that voter dapp has no regressions
- see that dapp uses new ancillary functions under the hood

### PR DESCRIPTION
**Pull request by daywiss**
Signed-off-by: David Adams <david@umaproject.org>

**Motivation**

#2371

**Summary**

Adds support for latest contract abi with ancillary data and fixes some bugs.

**Details**
- updates package file to latest abis
- uses ancillary data in calls when needed. also sets it to 0x since there is no ability to specify in this pr
- fixes infinite loop in requesting resolved requests
- tested with run_test script for 2key and non 2key scenarios
- "fixes" bug when decoding encrypted data which i ran into again, see #1969. this was resolved by updating the library version of `eth-crypto`

**Issue(s)**

Fixes #2371
Fixes #1969

### CODE REVIEW

File: packages/voter-dapp/src/ResolvedRequests.js
Comment (daywiss) [inline packages/voter-dapp/src/ResolvedRequests.js:87]: This is what was causing infinite loop

File: packages/voter-dapp/src/ActiveRequests.js
Comment (daywiss) [inline packages/voter-dapp/src/ActiveRequests.js:254]: this is ugly but for some reason ev.returnValues.ancillaryData is undefined, even though it was sent as 0x on the vote. This causes an issue if you dont default it to '0x'

File: packages/core/scripts/mainnet/ProposeAdmin.js
Comment (nicholaspai) [inline packages/core/scripts/mainnet/ProposeAdmin.js:35]: +1

File: packages/voter-dapp/run_tests.sh
Comment (nicholaspai) [inline packages/voter-dapp/run_tests.sh:181]: Thanks for updating this script!

File: packages/voter-dapp/src/ActiveRequests.js
Comment (nicholaspai) [inline packages/voter-dapp/src/ActiveRequests.js:79]: `toPriceRequestKey` and `toVotingAccountAndPriceRequestKey` now implement the same function. I find it a bit more readable to define the different vars that will be joined with commas, similar to how you define it in the comments. Why'd you change it?

File: packages/voter-dapp/src/ActiveRequests.js
Comment (nicholaspai) [reply packages/voter-dapp/src/ActiveRequests.js:254]: I experienced this also when testing the `OptimisticOracleClient` and I think using the `var || 0x` is better than how I addressed it which was the ternary operator: `var ? var : 0x`

File: packages/voter-dapp/src/ActiveRequests.js
Comment (nicholaspai) [inline packages/voter-dapp/src/ActiveRequests.js:323]: Does this still fail since we last talked? Catching this error might confuse users in case the messages cannot be decrypted, but I suppose its better than crashing the App.

File: packages/voter-dapp/src/ActiveRequests.js
Comment (nicholaspai) [inline packages/voter-dapp/src/ActiveRequests.js:367]: Shouldn't we be reading the ancillary data from `decryptedCommits` or is the decryption not working?

Comment (nicholaspai) [reply packages/voter-dapp/src/ActiveRequests.js:367]: Actually, see my comment below in the `commits.push` code.

File: packages/voter-dapp/src/ActiveRequests.js
Comment (nicholaspai) [inline packages/voter-dapp/src/ActiveRequests.js:401]: So clearly the committed `ancillaryData` and the revealed `ancillaryData` have to be identical. Furthermore, in the future we might want to change this field dynamically depending on the type of vote we're having. Therefore I think its sensible to set a constant ANCILLARY_DATA object at the file level that both this and reveals can read from. This way we can more easily upgrade the ancillary data logic to be dynamic based on types of votes.

File: packages/voter-dapp/src/ActiveRequests.js
Comment (nicholaspai) [inline packages/voter-dapp/src/ActiveRequests.js:424]: Wait, why do you use `pendingRequests[index].ancillaryData` here but hardcode `0x` above?

File: packages/voter-dapp/src/ResolvedRequests.js
Comment (nicholaspai) [reply packages/voter-dapp/src/ResolvedRequests.js:87]: Why does this fix the infinite loop? Really elegant solution if so

**Review by nicholaspai — COMMENTED** [review]
The changes are well implemented and overall its not much code which is awesome! I have a comment about how you set/get `ancillaryData` for commits and reveals.

Also, if this PR goes in will the voter dApp no longer be compatible with the current DVM? If so, we should discuss how to handle votes until we migrate the DVM. Perhaps we deploy a version of the web app to a new URL.

File: packages/voter-dapp/src/ActiveRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ActiveRequests.js:79]: Yea good point, i was a bit on the fence about this. The reason why is that it can accept arbitrary arguments now, for instance with or without ancillary data if needed, though this might not be a good enough reason.  I dont mind changing it back.

File: packages/voter-dapp/src/ActiveRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ActiveRequests.js:323]: This was failing for me until I bumped the eth-crypto version from 1.6.0 to 1.7.0 in the common package. I dont know if this will break other things, but doesnt seem like it so far.  I left the try catch in since this wasnt really being handled.

File: packages/voter-dapp/src/ActiveRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ActiveRequests.js:367]: you might be right about this, probably a mistake.

File: packages/voter-dapp/src/ActiveRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ActiveRequests.js:424]: Yea I think i made some mistakes here, good catch

File: packages/voter-dapp/src/ActiveRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ActiveRequests.js:401]: The next PR will be to add a field so the user can change ancillary data, but yea the default should at least be a constant.

File: packages/voter-dapp/src/ResolvedRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ResolvedRequests.js:87]: i think because if you dont unclude the `[]` it will run every render, and its probably causing a rerender when it runs. by adding the `[]` it means just run once.

**Comment by daywiss** [comment]
@nicholaspai thanks for the review. I addressed your concerns. I'm still not 100% sure of this but it does work correctly in the test. I think the next addition when adding user input for ancillary data should shake out further issues.

**Comment by daywiss** [comment]
> The changes are well implemented and overall its not much code which is awesome! I have a comment about how you set/get `ancillaryData` for commits and reveals.
> 
> Also, if this PR goes in will the voter dApp no longer be compatible with the current DVM? If so, we should discuss how to handle votes until we migrate the DVM. Perhaps we deploy a version of the web app to a new URL.

I think @mrice32 mentioned he has a plan for this release and noone seemed too concerned about this... Maybe matt can comment

File: packages/voter-dapp/src/ResolvedRequests.js
Comment (nicholaspai) [reply packages/voter-dapp/src/ResolvedRequests.js:87]: I think the second arg to any of the `useX` hooks is an array of dependencies, so I guess if you leave out the array of dependency then it will always re-render. If you explicitly set the dependencies to non with `[]` then it should only render once.

File: packages/voter-dapp/src/ActiveRequests.js
Comment (nicholaspai) [reply packages/voter-dapp/src/ActiveRequests.js]: > The reason why is that it can accept arbitrary arguments now, for instance with or without ancillary data if needed

This is definitely useful, but I think its confusing to have two functions implement the same logic. So if you want the flexibility of an unspecified amount of args, maybe consolidate the functions into one `toPriceRequestKey`

File: packages/voter-dapp/src/ActiveRequests.js
Comment (nicholaspai) [reply packages/voter-dapp/src/ActiveRequests.js:329]: Got it. I'm OK with leaving that here so the App doesn't crash, but its still going to be jarring for the user if they can't see their committed value (they will wonder, "did my commit go through correctly"?)

File: packages/voter-dapp/src/ResolvedRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ResolvedRequests.js:87]: yes thats my understanding and explained a lot better

File: packages/voter-dapp/src/ActiveRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ActiveRequests.js:329]: yea thats a great point but im not aware of any infrastructure within the voter dapp to alert the user to errors. if you have a suggestion on how to do this LMK

**Comment by mrice32** [comment]
> > The changes are well implemented and overall its not much code which is awesome! I have a comment about how you set/get `ancillaryData` for commits and reveals.
> > Also, if this PR goes in will the voter dApp no longer be compatible with the current DVM? If so, we should discuss how to handle votes until we migrate the DVM. Perhaps we deploy a version of the web app to a new URL.
> 
> I think @mrice32 mentioned he has a plan for this release and noone seemed too concerned about this... Maybe matt can comment

@daywiss I think the simplest thing to do is just wait until the upgrade is over to merge this.

**Comment by pemulis** [comment]
While we wait on the upgrade to merge this, should we resolve the merge conflicts?

**Comment by daywiss** [comment]
> While we wait on the upgrade to merge this, should we resolve the merge conflicts?

Ill fix conflicts when we are ready to go with this pr. shouldnt take long

File: packages/core/scripts/local/RequestOraclePrice.js
Comment (daywiss) [inline packages/core/scripts/local/RequestOraclePrice.js:70]: this script asks for new prices. im allowing this script to run either function to ensure when testing voter dapp both situations work ( and it does)

File: packages/voter-dapp/src/ActiveRequests.js
Comment (daywiss) [inline packages/voter-dapp/src/ActiveRequests.js:144]: modified this due to build warnings. most warnings are gone now

File: packages/voter-dapp/src/DesignatedVotingDeployment.js
Comment (daywiss) [inline packages/voter-dapp/src/DesignatedVotingDeployment.js:58]: this was added to stop compiler warning

File: packages/voter-dapp/src/ResolvedRequests.js
Comment (daywiss) [inline packages/voter-dapp/src/ResolvedRequests.js:155]: again recommended change by compiler

File: packages/voter-dapp/package.json
Comment (nicholaspai) [inline packages/voter-dapp/package.json:12]: What did this binding do before?

File: packages/voter-dapp/run_tests.sh
Comment (nicholaspai) [inline packages/voter-dapp/run_tests.sh:42]: Great to see this complexity removed.

File: packages/voter-dapp/src/ActiveRequests.js
Comment (nicholaspai) [reply packages/voter-dapp/src/ActiveRequests.js:144]: Super strange that this `useEffect()` hook would depend on the value of `hexToUtf8`, which is a function?

File: packages/voter-dapp/src/ActiveRequests.js
Comment (nicholaspai) [inline packages/voter-dapp/src/ActiveRequests.js:443]: Nice adding this here, not sure how you would have thought to do so without remembering about this button but +1

File: packages/voter-dapp/src/DesignatedVotingDeployment.js
Comment (nicholaspai) [reply packages/voter-dapp/src/DesignatedVotingDeployment.js:58]: Why does the `rel` fix the compiler warning OOC?

File: packages/voter-dapp/src/ResolvedRequests.js
Comment (nicholaspai) [inline packages/voter-dapp/src/ResolvedRequests.js:135]: +1

Comment (nicholaspai) [reply packages/voter-dapp/src/ResolvedRequests.js:135]: Can you not do `allResolvedEvents?.length`?

File: packages/voter-dapp/src/ResolvedRequests.js
Comment (nicholaspai) [reply packages/voter-dapp/src/ResolvedRequests.js:155]: Makes no sense...

File: packages/voter-dapp/src/RetrieveRewards.js
Comment (nicholaspai) [inline packages/voter-dapp/src/RetrieveRewards.js:235]: Also super strange to see `web3.utils` here

**Review by nicholaspai — APPROVED** [review]
Really clean code and its great to see a bunch of hack fixes going away!

File: packages/voter-dapp/package.json
Comment (daywiss) [reply packages/voter-dapp/package.json:12]: this was due to @mrice32 PR #2586. allowed us to load in new and old version of contract into app so we had both abis avaialble, or something like that

File: packages/voter-dapp/src/ActiveRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ActiveRequests.js:443]: @nicholaspai actually your run_tests script is very thorough and once i added requests with both ancillary and non ancillary data I had to do a lot more updates. pretty much every change in this pr was found thanks to that script 👍

File: packages/voter-dapp/src/ResolvedRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ResolvedRequests.js:135]: interesting, yea i havent used this new syntax yet but its a good candidate for it. can update

File: packages/voter-dapp/src/RetrieveRewards.js
Comment (daywiss) [reply packages/voter-dapp/src/RetrieveRewards.js:235]: yea... i was just following the compiler warnings here, i dont really understand why it wanted this

File: packages/common/package.json
Comment (daywiss) [inline packages/common/package.json:45]: not sure if this is strictly necessary, but running locally without this change doesnt seem to be an option

File: packages/voter-dapp/src/DesignatedVotingDeployment.js
Comment (daywiss) [reply packages/voter-dapp/src/DesignatedVotingDeployment.js:58]: image

File: packages/voter-dapp/src/ResolvedRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ResolvedRequests.js:155]: image

File: packages/voter-dapp/src/ActiveRequests.js
Comment (mrice32) [inline packages/voter-dapp/src/ActiveRequests.js:257]: Is ev.returnValues.ancillaryData null when the bytes field is empty? I would expect it to be `0x`...

File: packages/voter-dapp/src/ActiveRequests.js
Comment (mrice32) [inline packages/voter-dapp/src/ActiveRequests.js:332]: If we're going to catch this error, can we return a string value that indicates the error? I think this should be distinguishable from no vote in some way. Something like "ERROR: DECRYPTION FAILED"? Would be ideal to show that in the "value" field in the dapp. Another (slightly less ideal) solution could be to print out the error using console.error so we can diagnose the issue in prod. Otherwise the error gets swallowed and there's no way to find the cause.

File: packages/voter-dapp/src/ActiveRequests.js
Comment (mrice32) [inline packages/voter-dapp/src/ActiveRequests.js:364]: commented code?

File: packages/voter-dapp/src/ResolvedRequests.js
Comment (mrice32) [reply packages/voter-dapp/src/ResolvedRequests.js:155]: The second part confuses me @daywiss -- can you shed a light on why the blacklist "doesn't re-render the component"?

**Review by mrice32 — APPROVED** [review]
LGTM!

File: packages/voter-dapp/src/ActiveRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ActiveRequests.js:144]: its not that it doesnt depend on it, but react warns that mutations to `IDENTIFIER_BLACKLIST` will not re-render component, so it isnt needed
image

File: packages/voter-dapp/src/ResolvedRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ResolvedRequests.js:155]: im guessing its because this is state defined outside of a react component, so react cant react to changes to this variable. on top of that its a constant as well, so its kind of moot, it should never change.

File: packages/core/scripts/local/RequestOraclePrice.js
Comment (daywiss) [reply packages/core/scripts/local/RequestOraclePrice.js:70]: depends on what you mean by does it work. I've updated the test script to create price requests that hit both of these branches, and it works as expected: price requests are made with and without ancillary data. Its possible that line 70 is still calling the ancillary version of the function though, which could be giving me similar behavior.

File: packages/voter-dapp/src/ActiveRequests.js
Comment (daywiss) [reply packages/voter-dapp/src/ActiveRequests.js:332]: yes, this should log to console.error since theres no way to really notifiy the user in this app

**Pull request merged**
